### DISCUSSION TITLE
FreeIPA <- Keycloak

### DISCUSSION THREAD

**Discussion by k1ck3r** [keycloak/keycloak#13691] [Q&A]
Hello all,

I have an issue configuring both systems Keycloak and FreeIPA to work with User Federation. Configuration on Keycloak side for the ldap (FreeIPA server) is as follows:

- LDAPs configuration
- Keytab from FreeIPA generated with admin user

The below screenshot is from the Keycloak User Federation:
![image](https://user-images.githubusercontent.com/32503925/183832732-d3a2130a-5fbe-44e0-8fce-358c006d4808.png)
![image](https://user-images.githubusercontent.com/32503925/183832785-3677737b-18cb-45ee-8764-b441381487ab.png)

Importing users works flawlessly but the problems comes when I try to create user in Keycloak and expect it to be created on FreeIPA side - WRITABLE is on, and keycloak machine is enrolled into FreeIPA as a client (both OSes are Alma). There is no error, and Keycloak indicates that a new user is created.

However, in FreeIPA's web interface the user is missing and the most frustrating thing is if i try to create the very same username, FreeIPA returns that it can't add the user, because it already exists. I guess the issue would be somewhere either in Username/RDN LDAP attribute or UUID or even Custom User LDAP filter, but i'm lost a bit.

In case someone wants to help here what i've tried to play with:

- Setting UUID Ldap attribute to ipaUniqueID, but using it, returns 0 user when trying to sync, and creating user from Keycloak returns error
- Setting custom ldap filter to match a group from the LDAP - no binding with admin user could be achieved, thus no user could be synced

Anyhelp on this will be much appreciated :") 

Thank you in advance

**Comment by Avsynthe** [comment]
Hey mate!

Did you end up figuring this out? It seems that FreeIPA is basically locked into having to manually create users in it and there are absolutely no self-service solutions that do more than import from it and write password changes to it. 

I came across these two unanswered threads here:

https://keycloak.discourse.group/t/error-when-syncing-keycloak-with-freeipa/4183

https://serverfault.com/questions/1049936/unable-to-add-users-from-keycloack-on-freeipa-via-ldap-although-keycloack-host-h

In a nutshell, it seems like it simply can't be done, which is INSANE as both FreeIPA and Keycloak are developed by Red Had. Not to mention this exact function works with any other connected LDAP source other than FreeIPA

I also made a reddit post here:

https://www.reddit.com/r/selfhosted/comments/yqkgf8/can_authentikkeycloak_andor_any_other_auth/

I think I'm about to migrate everything to Authentik and run an LDAP server there, which is insane considering the implication that I'm about to move from enterprise level software to a solution that was created by one guy in their spare time, that only just became a company and is only now just hiring for it.

Craziness

**Comment by Macr0Nerd** [comment]
I want to provide some insight here. In order for this to work, (I presume) that you have to create a mapper for all the ldap attributes and object classes that FreeIPA expects to be there. If you've looked at the ldap information for a newly created user, you can see it's easily 40+ attributes that need to be set. Additionally, if you're using the `ipaUniqueID` field at all (as Nextcloud does by default when users are provisioned by LDAP), then you may not be able to do so at all. I suppose in theory it may be possible to create this value yourself, but I am unsure how FreeIPA will react to this. Because FreeIPA handles so much - more than what is typically configured in Keycloak I presume) there isn't really an easy way around this without essentially doing the work to create a user in LDAP itself.

I don't really see this as an issue in FreeIPA but moreso an issue between an IdM and an IdP. The IdM and IdP each serve different roles, and if you want to propogate from the IdP -> IdM I presume you need to put in all the legwork that the IdM would do for you. For myself, I will continue to create users in FreeIPA and import them into Keycloak, but I may investigate what is necessary to create users in Keycloak and if it is indeed possible.

Apologies for the necrobump, I came across this discussion while researching the issue myself and wanted to leave behind what I've figured out for the next person.
